### PR TITLE
[Perf] Optimize Qwen3-Omni predictor hot path

### PR DESCRIPTION
**Pull request by luojiaxuan**
## Summary

Optimize the Qwen3-Omni talker code predictor hot path with larger Triton fusion instead of small copy-only fusion.

Changes:
- Add optional Triton kernels for predictor input staging and code sampling/staging.
- Fuse `argmax logits -> code write -> embedding lookup -> predictor input write -> summed embedding update`.
- Fuse initial layer0 code embedding and predictor input staging.
- Use SDPA `enable_gqa=True` to avoid explicit KV repeat when available.
- Add CUDA unit coverage for the fused predictor kernels, including strided views and argmax tie behavior.

## Validation

Unit checks:
- `python3 -m py_compile sglang_omni/models/qwen3_omni/components/predictor_kernels.py sglang_omni/models/qwen3_omni/components/talker.py tests/unit_test/qwen3_omni/test_predictor_kernels.py`
- H100: `pytest -q tests/unit_test/qwen3_omni/test_predictor_kernels.py` -> 2 passed
- H100: `pytest -q tests/unit_test/qwen3_omni/test_talker.py` -> 63 passed

H100 benchmark, Qwen3-Omni SeedTTS EN full set, `temperature=0`, `max_concurrency=16`, `max_samples=1088`, voice clone enabled:

| Metric | Baseline | Patch | Change |
| --- | ---: | ---: | ---: |
| Completed | 1088/1088 | 1088/1088 | same |
| Throughput | 5.576 req/s | 5.696 req/s | +2.15% |
| Audio throughput | 20.226 audio-s/s | 20.507 audio-s/s | +1.39% |
| Output throughput | 79.5 tok/s | 81.2 tok/s | +2.14% |
| Latency mean | 2.864s | 2.805s | -2.06% |
| Latency p95 | 4.250s | 4.204s | -1.08% |
| Latency p99 | 5.100s | 4.731s | -7.24% |
| RTF mean | 0.8073 | 0.7965 | -1.34% |
| RTF p99 | 1.4234 | 1.3356 | -6.17% |
| WER corpus | 2.27% | 2.05% | -0.22 pp |
| >50% WER outliers | 5 | 2 | better |

Result dirs:
- Baseline: `/data/results/qwen3_omni_main_seedtts1088_c16_t0_gpu4_v1`
- Patch: `/data/results/qwen3_omni_large_fusion_seedtts1088_c16_t0_gpu4_v1`

Kernel microbench on H100:
- batch=16: stage fusion 1.85x, sample/stage fusion 1.89x
- batch=32: stage fusion 1.85x, sample/stage fusion 1.88x

## Notes

This targets the predictor staging/sample hot path. End-to-end speedup is modest on the full benchmark, suggesting the remaining Qwen3-Omni bottleneck is still broader scheduling/host/code2wav/runtime overhead rather than only this predictor fragment.

### CODE REVIEW

**Review by luojiaxuan — COMMENTED** [review]
Reviewed and pushed 9d004b1 to remove a duplicate ready-request scan in code2wav batching. Checks: py_compile for touched Qwen files and git diff --check.

**Comment by luojiaxuan** [comment]
Full H200 benchmark done with the 30B model downloaded into the shared HF cache.

Setup:
- Host/container: hyper00 H200, `sglang-omni-jaxan-pr872-bench-0626`
- Shared mounts: `/data01/cache/huggingface -> /root/.cache/huggingface`, `/data02/jaxan -> /data`
- Model snapshots: `Qwen/Qwen3-Omni-30B-A3B-Instruct`, `Qwen/Qwen3-ASR-1.7B`
- Baseline: `d85131f`; fused experiment: `17f9bfe`; current PR head: `e55e75c`

How S2S stress was run:
- Script: `python -m benchmarks.eval.benchmark_omni_rollout_stress`
- Config: same MMMU sample duplicated N times, `--rollout-counts 1,2,4,8,16`, `--max-tokens 256`, `--temperature 0`, audio enabled, `stream=false`
- It measures same-prompt concurrent rollout throughput/latency and writes profiler events; it is not a streaming first-audio benchmark.

Rejected direction: fused `talker_ar -> code2wav` same-process topology.
- It improved S2S stress a lot, but full SeedTTS throughput did not meet the default-on gate and streaming TTFT regressed badly.
- Kept it opt-in only via `SGLANG_OMNI_QWEN3_FUSE_TALKER_CODE2WAV=1` or `examples/configs/qwen3_omni_colocated_h200_fused.yaml`.

Current direction: non-fused topology + streaming-first code2wav window.
- Added first streaming decode window for code2wav with default size 8, then subsequent windows keep the normal size 10.
- Non-streaming requests keep the normal first window, so full SeedTTS/S2S batching semantics are not changed by this knob.
- Kill switch: `SGLANG_OMNI_QWEN3_CODE2WAV_FIRST_CHUNK_SIZE=0`.

Current full SeedTTS EN VC 1088, concurrency 16, temp 0:

| Metric | Baseline | Current | Delta |
| --- | ---: | ---: | ---: |
| Throughput | 5.560 req/s | 5.202 req/s | -6.44% |
| Audio throughput | 20.028 s/s | 20.224 s/s | +0.98% |
| Mean RTF | 0.8326 | 0.7931 | -4.74% |
| Mean latency | 2.870s | 2.940s | +2.44% |
| p95 latency | 4.287s | 4.115s | -4.01% |
| p99 latency | 7.450s | 5.073s | -31.91% |
| Corpus WER | 2.1268% | 2.1865% | +0.0597 pp |

Current streaming TTFT (`FIRST_CHUNK_SIZE=8`):

| Prompt | Baseline mean | Current mean | Delta | Baseline p95 | Current p95 | Delta |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| short | 0.304s | 0.294s | -3.14% | 0.355s | 0.344s | -3.22% |
| medium | 0.355s | 0.334s | -5.88% | 0.396s | 0.337s | -15.00% |

Current S2S stress, all counts succeeded:

| Rollouts | Baseline qps | Current qps | Delta | Baseline p95 | Current p95 | Delta |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 0.071 | 0.082 | +15.49% | 14.159s | 12.223s | -13.67% |
| 2 | 0.151 | 0.153 | +1.32% | 13.193s | 13.011s | -1.38% |
| 4 | 0.225 | 0.244 | +8.44% | 17.759s | 16.226s | -8.63% |
| 8 | 0.313 | 0.281 | -10.22% | 25.382s | 26.578s | +4.71% |
| 16 | 0.397 | 0.411 | +3.53% | 39.656s | 37.744s | -4.82% |

Validation for the latest change:
- local `python3 -m py_compile` and `git diff --check`
- container: `pytest -q tests/unit_test/qwen3_omni/test_code2wav.py` -> 10 passed

Net: the fused topology was the wrong default direction. Current head prioritizes full SeedTTS/TTFT safety; S2S N=8 is still not a win without the fused topology, so I am not claiming that gate as passed.

**Comment by luojiaxuan** [comment]
Updated after rerunning Video-AMME with video decode cache disabled. The previous cache-on Video-AMME numbers were inflated by repeating the same ci-50 samples to 150 requests, so I no longer count them as real-request E2E throughput evidence.

Definitions: baseline = pre-E2E-optimization head `9d004b1`; candidate = final pushed PR head `c566ce5`. Video decode cache is now opt-in (`SGLANG_OMNI_VIDEO_DECODE_CACHE=1`) and is not part of the default benchmark table below.

Implementation summary:
- Used #894 Video-AMME Talker stress from a separate checkout; the harness is not mixed into this PR.
- Kept fused talker/code2wav opt-in, not default.
- Default colocated speech path keeps partial-start on.
- Code2wav defaults to wait=5ms and non-stream chunk size=20.
- Video decode cache is opt-in only; it can help repeated local-video workloads, but is not counted as default Video-AMME E2E gain.
- Serialized same-GPU colocated stage startup to reduce startup contention.

Video-AMME Talker closed-loop, ci-50 repeated to 150 requests, request-rate=inf, video decode cache OFF:

| Stage | Completed / audio | QPS | QPS delta | p95 latency | p95 delta | Queue p95 | Accuracy |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| c8 | 150/150, 100% | 1.196 -> 1.250 | +4.52% | 10.607 -> 10.266s | -3.21% | 115.555 -> 109.332s | 0.6733 -> 0.6800 |
| c16 | 150/150, 100% | 1.533 -> 1.595 | +4.04% | 17.816 -> 15.252s | -14.39% | 84.717 -> 79.834s | 0.6667 -> 0.6733 |
| c32 | 150/150, 100% | 1.580 -> 1.704 | +7.85% | 28.716 -> 27.172s | -5.38% | 75.568 -> 72.579s | 0.6733 -> 0.6800 |

Video-AMME conclusion: latency is positive, c32 throughput is positive, but c16 throughput is +4.04%, below the original +5% gate. Therefore I do not claim the default path fully passes the Video-AMME throughput gate.

Video-AMME sustained c16, request-rate=1.0, video decode cache OFF:

| Metric | Baseline | Candidate | Delta |
| --- | ---: | ---: | ---: |
| QPS | 0.977 | 0.982 | +0.51% |
| p95 latency | 7.630s | 6.555s | -14.09% |
| audio throughput | 13.357 | 13.694 | +2.52% |
| queue p95 | 0.000s | 0.000s | 0.00s |
| accuracy | 0.7000 | 0.7000 | +0.00 pp |

SeedTTS full 1088:

| Metric | Baseline | Candidate | Delta |
| --- | ---: | ---: | ---: |
| QPS | 6.063 | 6.458 | +6.51% |
| audio throughput | 21.881 | 23.722 | +8.41% |
| RTF mean | 0.7490 | 0.6907 | -7.78% |
| p95 latency | 3.923s | 3.668s | -6.50% |
| p99 latency | 4.622s | 4.475s | -3.18% |
| WER | 1.9007% | 1.8588% | -0.0419 pp |

Streaming TTFT, 50 repeats:

| Prompt | Metric | Baseline | Candidate | Delta |
| --- | ---: | ---: | ---: | ---: |
| short | TTFT p95 | 0.345856s | 0.333680s | -3.52% |
| short | total latency p95 | 0.905597s | 0.880896s | -2.73% |
| medium | TTFT p95 | 0.352105s | 0.317699s | -9.77% |
| medium | total latency p95 | 1.164149s | 1.130003s | -2.93% |

Tests:
- `python3 -m py_compile` on changed files passed locally and in the H200 container.
- Required qwen3_omni tests passed: 46.
- qwen3_omni config/CLI/example/pipeline tests passed: 181.
- #894 Video-AMME harness unit test passed: 7.
- Video preprocessing + startup batching targeted tests passed: 5.
- After making video decode cache opt-in, `pytest -q tests/unit_test/preprocessing/test_video.py` passed 5.

Artifacts on hyper01:
- No-cache Video-AMME: `/data/pr872-fullbench/runs/20260627-videoamme-e2e-v16`
- SeedTTS full: `/data/pr872-fullbench/runs/20260627-seedtts-full-nonstream20-v1`
- Streaming TTFT: `/data/pr872-fullbench/runs/20260627-streaming-ttft-nonstream20-v2`

File: sglang_omni/preprocessing/video.py [hunk 1]
```diff
@@ -23,11 +27,125 @@
...
+            _video_decode_cache_bytes -= old.nbytes
+        while _video_decode_cache and _video_decode_cache_bytes + nbytes > max_bytes:
+            _, evicted = _video_decode_cache.popitem(last=False)
+            _video_decode_cache_bytes -= evicted.nbytes
+        _video_decode_cache[key] = _VideoDecodeCacheEntry(
+            video=video,
+            sample_fps=sample_fps,
+            nbytes=nbytes,
+        )
+        _video_decode_cache_bytes += nbytes
```
[the remark this answers is not in the document]
Comment (luojiaxuan) [reply sglang_omni/preprocessing/video.py]: Fixed by removing the global byte counter and computing cache size from entries during insertion.
